File: TheatreJS/ModelPhysik/hautmaske.py
# -*- coding: utf-8 -*-
u"""Hautmaske fuer den Server-Videoweg — dieselben Regeln wie im Browser.

Der Browser (`static/viewer/gemeinsam/hautmaske.js`, 11.09.2026) zeichnet
die Haut unter dem Stoff nicht mehr: Was nicht gezeichnet wird, kann in
keiner Pose durchkommen. Das Video vom Server rendert seinen eigenen
Koerper, und dort galt die Maske nicht — die Haut kam durch die Leggings,
obwohl die Szene sauber war. Hier steht die Rechnung in NumPy, Regel fuer
Regel gleich (`core/tests/unit/test_hautmaske_python.py` haelt beide
Fassungen am selben Kunstkoerper aneinander):

* VERDECKT ist ein Koerperpunkt, wenn der Strahl entlang seiner Normale ein
  Stoffdreieck trifft — hoechstens `ABSTAND_M` darueber, hoechstens
  `TIEFE_M` in der Haut. Gesucht wird bei den Dreiecken bis zwei Ringe um
  den naechsten Stoffpunkt.
* Ein Randstreifen (`RANDRINGE`) bleibt nur an LOCKEREN Stoffkanten frei
  (mehr als `ENG_M` ueber der Haut); eine anliegende Kante maskiert bis zum
  Rand — der Bund der 2-mm-Leggings war sonst genau die Haut, die herauskam.
* Freie Inseln im Verdeckten (bis `INSEL_MAX` Punkte) gelten als verdeckt
  (Achselfalte: die Normale trifft dort keinen Stoff).
* Ein Dreieck faellt weg, wenn alle drei Ecken verdeckt sind; die
  verdeckten Ecken der Randdreiecke ziehen sich beim Rendern `EINZUG_M`
  nach innen (`Filmrender`).
* Stoff unter Stoff (`lagenmaske.py`): Wer aussen liegt, wird gezaehlt; die
  Normalen kommen von der Haut, nie vom Stoffnetz.

Seit dem 12.09.2026 je Klasse eine Datei: `maskengeometrie.py` (Normalen,
Strahl), `maskeninseln.py`, `lagenmaske.py`.
"""
import numpy as np
from scipy.sparse import coo_matrix, csr_matrix, identity
from scipy.spatial import cKDTree

from maskengeometrie import Geometrie
from maskeninseln import Maskeninseln


class Hautmaske:
    u"""Welche Koerperpunkte unter einem Stueck liegen."""

    ABSTAND_M = 0.025
    TIEFE_M = 0.005
    RANDRINGE = 2
    ENG_M = 0.004
    INSEL_MAX = 200
    EINZUG_M = 0.010

    @classmethod
    def verdeckt(cls, koerper, dreiecke, stoffe, abstand=None, tiefe=None,
                 randringe=None, eng=None, normalen=None, suchweite=None,
                 inseln=None):
        u"""Bool je Koerperpunkt. `stoffe`: Liste von (punkte, dreiecke)."""
        w = cls.werte(abstand=abstand, tiefe=tiefe, randringe=randringe,
                      eng=eng, suchweite=suchweite, inseln=inseln)
        koerper = np.asarray(koerper, dtype=np.float64)
        if normalen is None:
            normalen = Geometrie.normalen(koerper, dreiecke)
        normalen = np.asarray(normalen, dtype=np.float64)
        baum = cKDTree(koerper)
        maske = np.zeros(len(koerper), dtype=bool)
        for punkte, tris in stoffe:
            if cls._leer(punkte, tris):
                continue
            cls._ein_stueck(koerper, normalen, baum, maske,
                            np.asarray(punkte, dtype=np.float64),
                            np.asarray(tris, dtype=np.int64).reshape(-1, 3),
                            w['abstand'], w['tiefe'], w['randringe'], w['eng'],
                            w['suchweite'])
        if w['inseln'] > 0 and dreiecke is not None:
            Maskeninseln.schliessen(maske, dreiecke, w['inseln'])
        return maske

    @classmethod
    def werte(cls, **wahl):
        u"""Die Masse der Maske: `None` heisst Klassenkonstante; die Suchweite
        folgt ohne Angabe dem groesseren von Abstand und Tiefe."""
        vorgaben = {'abstand': cls.ABSTAND_M, 'tiefe': cls.TIEFE_M,
                    'randringe': cls.RANDRINGE, 'eng': cls.ENG_M,
                    'inseln': cls.INSEL_MAX}
        w = {name: (vorgabe if wahl.get(name) is None else wahl[name])
             for name, vorgabe in vorgaben.items()}
        suchweite = wahl.get('suchweite')
        w['suchweite'] = (max(w['abstand'], w['tiefe']) if suchweite is None
                          else suchweite)
        return w

    @staticmethod
    def _leer(punkte, tris):
        u"""Ein Stueck ohne Punkte oder ohne Dreiecke verdeckt nichts."""
        return (punkte is None or tris is None
                or not len(punkte) or not len(tris))

    @classmethod
    def _ein_stueck(cls, koerper, normalen, baum, maske, P, T,
                    abstand, tiefe, ringe, eng, suchweite):
        rand, nachbarn = cls.randpunkte(T, len(P))
        locker = cls.lockere_randpunkte(P, rand, koerper, normalen, baum, eng)
        gesperrt = cls.ringe_um(locker, nachbarn, ringe)
        ok = ~gesperrt[T].any(axis=1)
        umkreis = cls.dreiecke_im_umkreis(T, len(P), ok, nachbarn)
        d, j = cKDTree(P).query(koerper, distance_upper_bound=suchweite, workers=-1)
        kand = np.where(~maske & np.isfinite(d))[0]
        if not len(kand):
            return
        t = Geometrie.strahl_dreiecke(P, T, umkreis[j[kand]],
                                      koerper[kand], normalen[kand])
        with np.errstate(invalid='ignore'):
            treffer = np.any((t >= -tiefe) & (t <= abstand), axis=1)
        maske[kand[treffer]] = True

    # ---------------------------------------------------------------- Rand

    @staticmethod
    def randpunkte(T, n):
        u"""Bool je Punkt (an einer offenen Kante) und die Nachbarschaft
        als duenn besetzte (n, n)-Matrix."""
        kanten = np.vstack([T[:, [0, 1]], T[:, [1, 2]], T[:, [2, 0]]])
        kanten.sort(axis=1)
        schluessel = kanten[:, 0].astype(np.int64) * n + kanten[:, 1]
        einzig, zaehler = np.unique(schluessel, return_counts=True)
        offen = einzig[zaehler == 1]
        rand = np.zeros(n, dtype=bool)
        rand[offen // n] = True
        rand[offen % n] = True
        a, b = einzig // n, einzig % n
        nachbarn = coo_matrix((np.ones(2 * len(a), dtype=np.int8),
                               (np.concatenate([a, b]), np.concatenate([b, a]))),
                              shape=(n, n)).tocsr()
        return rand, nachbarn

    @staticmethod
    def lockere_randpunkte(P, rand, koerper, normalen, baum, eng):
        u"""Randpunkte, die mehr als `eng` ueber der Haut liegen (oder ohne
        Haut in 9 cm Umkreis)."""
        locker = np.zeros(len(P), dtype=bool)
        wo = np.where(rand)[0]
        if not len(wo):
            return locker
        d, i = baum.query(P[wo], distance_upper_bound=0.09, workers=-1)
        weit = ~np.isfinite(d)
        i = np.where(weit, 0, i)
        tief = np.einsum('ij,ij->i', P[wo] - koerper[i], normalen[i])
        locker[wo] = weit | (np.abs(tief) > eng)
        return locker

    @staticmethod
    def ringe_um(start, nachbarn, ringe):
        menge = start.astype(np.float64)
        for _ in range(int(ringe)):
            menge = np.clip(menge + nachbarn @ menge, 0, 1)
        return menge > 0

    @staticmethod
    def dreiecke_im_umkreis(T, n, ok, nachbarn):
        u"""(n, K)-Feld: je Stoffpunkt die zulaessigen Dreiecke bis zwei
        Ringe weit, mit -1 aufgefuellt."""
        nT = len(T)
        okT = np.where(ok)[0]
        if not len(okT):
            return np.full((n, 1), -1, dtype=np.int64)
        zeilen = np.concatenate([T[okT, 0], T[okT, 1], T[okT, 2]])
        spalten = np.concatenate([okT, okT, okT])
        B = csr_matrix((np.ones(len(zeilen), dtype=np.int8), (zeilen, spalten)),
                       shape=(n, nT))
        R = identity(n, dtype=np.int8, format='csr') + nachbarn
        R = R + R @ nachbarn
        C = (R @ B).tocsr()
        C.sum_duplicates()
        laengen = np.diff(C.indptr)
        K = max(int(laengen.max()), 1)
        aus = np.full((n, K), -1, dtype=np.int64)
        for i in range(n):
            s, e = C.indptr[i], C.indptr[i + 1]
            aus[i, :e - s] = C.indices[s:e]
        return aus

    # --------------------------------------------------------------- Index

    @staticmethod
    def index_ohne(dreiecke, maske):
        u"""Dreiecke ohne die, deren drei Ecken verdeckt sind."""
        T = np.asarray(dreiecke, dtype=np.int64).reshape(-1, 3)
        weg = maske[T].all(axis=1)
        return T[~weg], int(weg.sum())

    @classmethod
    def einzug(cls, punkte, dreiecke, maske, betrag=None):
        u"""Die verdeckten Punkte um `betrag` entlang ihrer Normale nach
        innen — fuer das Rendern der Randdreiecke."""
        betrag = cls.EINZUG_M if betrag is None else betrag
        aus = np.array(punkte, dtype=np.float64)
        N = Geometrie.normalen(aus, dreiecke)
        aus[maske] -= betrag * N[maske]
        return aus

File: static/viewer/gemeinsam/hautmaske.js
/**
 * Hautmaske — welche Körperpunkte unter einem Kleidungsstück liegen, und ein
 * Dreiecksindex ohne die verdeckten Flächen.
 *
 * BEFUND (Edgar, 11.09.2026, zweimal mit Bild: „bei einer animation kommt
 * der Körper durch die Kleidung hindurch", „natürlich ist es die Haut die
 * durch die Verformung durchkommt!!!"): Eine Leggings liegt 2 mm auf der
 * Haut, und in Dance1 stehen an Bund, Schritt und Knie weiße Splitter
 * heraus. Körperseitig gegen die Stoffdreiecke gemessen (Strahl entlang der
 * Körpernormale, Stoff hinten und keiner vorn): je Bild 3 bis 33 Körperpunkte
 * des Beckens vor dem Stoff, 2 bis 5 mm tief, am kreuzenden Knie 29 mm.
 * Gleiche Gewichte helfen nicht — mit den Hautgewichten der 70.851 Punkte
 * (nächster Punkt oder baryzentrisch über das nächste Dreieck) blieben es 16
 * und 31 statt 18 und 33. Zwei Flächen, die 2 mm auseinanderliegen und
 * getrennt gehäutet werden, kommen sich an jedem Gelenk um Millimeter nahe;
 * das ist keine Frage der Gewichte.
 *
 * DER WEG, DEN JEDE SPIELENGINE GEHT — und MakeHuman mit `delete_verts`
 * (08.09.2026, dort 23,5 % der Anzugpunkte in der Haut, ohne dass man es
 * sieht): Die Haut unter dem Stoff wird nicht gezeichnet. Was nicht gezeichnet
 * wird, kann nicht durchkommen — in keiner Pose, mit oder ohne Weichgewebe,
 * ohne Rechenzeit je Bild.
 *
 * VERDECKT ist ein Körperpunkt, wenn der Strahl entlang seiner Normale ein
 * Stoffdreieck trifft
 *   * höchstens `ABSTAND_M` über der Haut (25 mm — das Kontaktband der
 *     Durchstich-Messungen vom 07.09.2026, `KONTAKT_M`),
 *   * höchstens `TIEFE_M` IN der Haut (ein Stoff, der schon 5 mm drinsteckt,
 *     zählt noch als darüber — genau die Fälle, um die es geht),
 *   * und das Dreieck nicht in einem freien Randstreifen liegt — siehe unten.
 * Gesucht wird nur bei den Dreiecken um den nächsten Stoffpunkt (Gitter),
 * nicht über das ganze Stück.
 *
 * DER RANDSTREIFEN GILT NUR AN LOCKEREN KANTEN (Edgar, gleicher Abend, mit
 * Bild: „besser mit durchscheinen, aber noch immer nicht gefixt"). Die erste
 * Fassung ließ an JEDER offenen Stoffkante (Saum, Bund, Ausschnitt) zwei
 * Nachbarringe Haut stehen, damit ein Saum, der sich hebt, kein Loch
 * freigibt. Am Bund der 2-mm-Leggings war dieser Streifen (1,6 cm) genau
 * die Haut, die in Dance1 weiter herauskam — Bild 32, sechs Splitter entlang
 * der Bundkante, Beine und Knie sauber. Eine Kante, die auf der Haut liegt,
 * hebt sich nicht; eine, die `ENG_M` (4 mm) oder mehr Luft hat, kann es.
 * Deshalb wird je Randpunkt des Stoffs gemessen, wie weit er über der Haut
 * liegt (Abstand entlang der Körpernormale des nächsten Körperpunkts): Nur
 * von LOCKEREN Randpunkten aus bleiben `RANDRINGE` Ringe frei; an einer
 * anliegenden Kante reicht die Maske bis zum Rand. Der Preis dort ist ein
 * höchstens millimeterbreiter Spalt, wenn die Kante in Bewegung über die
 * Haut rutscht — an einem schwarzen Bund unsichtbar, gegen weiße Splitter
 * der bessere Tausch.
 *
 * Ein Dreieck des Körpers fällt aus dem Index, wenn ALLE drei Ecken verdeckt
 * sind; die Materialgruppen (Haut, Augen, Zähne, …) werden nachgezogen —
 * ohne das läge das Material des Innenmunds über dem Bauch (dieselbe Falle
 * wie bei UMA am 08.09.2026, `addGroup` zählt Indexeinträge).
 *
 * OHNE THREE.JS UND OHNE DOM — damit die Rechnung in Node prüfbar ist
 * (`core/tests/unit/test_js_hautmaske.py`). Die Geometrie-Bausteine stehen
 * in `hautmaskegeometrie.js`, die Three.js-Seite in `scene/hautverdeckung.js`,
 * Stoff über Stoff in `lagenmaske.js`.
 */
import { Hautmaskegeometrie as G } from './hautmaskegeometrie.js';
import { Maskeninseln } from './maskeninseln.js';

export class Hautmaske {

    /** Näher als das gilt als „unter dem Stoff" (Meter). */
    static ABSTAND_M = 0.025;
    /** So weit darf der Stoff schon IN der Haut stecken und zählt noch. */
    static TIEFE_M = 0.005;
    /** Nachbarringe um LOCKERE offene Stoffkanten, die frei bleiben. */
    static RANDRINGE = 2;
    /** Bis zu diesem Abstand über der Haut liegt eine Stoffkante AN. */
    static ENG_M = 0.004;
    /** Freie Inseln im Verdeckten bis zu dieser Punktzahl gelten als verdeckt. */
    static INSEL_MAX = 200;

    /**
     * Je Körperpunkt 1, wenn ihn eines der Stücke verdeckt.
     *
     * @param koerper      Float32Array/Float64Array, Punkte in Ruhelage (xyz)
     * @param dreiecke     Index des Körpers (drei Einträge je Dreieck)
     * @param stoffe       [{punkte, dreiecke}] — jedes Stück in derselben Lage
     * @param optionen     {abstand, tiefe, randringe, eng, normalen, suchweite, inseln}
     *                     `normalen`: fertige Punktnormalen des Körpers
     *                     (die Lagenmaske gibt einem Stoff die der Haut mit);
     *                     `suchweite`: Umkreis für den nächsten Stoffpunkt
     *                     (sonst das Größere von `abstand` und `tiefe`) —
     *                     Messungen fragen damit auch nach Stoff HINTER der
     *                     Haut (`abstand` negativ); `inseln`: Höchstgröße
     *                     freier Inseln, die noch geschlossen werden (0 = aus;
     *                     Messungen lassen die Maske roh).
     */
    static verdeckt(koerper, dreiecke, stoffe, optionen = {}) {
        const abstand = optionen.abstand ?? Hautmaske.ABSTAND_M;
        const tiefe = optionen.tiefe ?? Hautmaske.TIEFE_M;
        const ringe = optionen.randringe ?? Hautmaske.RANDRINGE;
        const eng = optionen.eng ?? Hautmaske.ENG_M;
        const suchweite = optionen.suchweite ?? Math.max(abstand, tiefe);
        const inseln = optionen.inseln ?? Hautmaske.INSEL_MAX;
        const n = koerper.length / 3;
        const maske = new Uint8Array(n);
        const normalen = optionen.normalen || G.normalen(koerper, dreiecke);
        const koerpergitter = G.punktgitter(koerper, G.ZELLE_M);
        const basis = { koerper, normalen, koerpergitter };
        for (const stoff of stoffe || []) {
            if (!stoff?.punkte?.length || !stoff?.dreiecke?.length) continue;
            Hautmaske._einStueck(basis, maske, stoff, abstand, tiefe, ringe, eng, suchweite);
        }
        if (inseln > 0 && dreiecke) Maskeninseln.schliessen(maske, dreiecke, inseln);
        return maske;
    }

    static _einStueck(basis, maske, stoff, abstand, tiefe, ringe, eng, suchweite) {
        const { koerper, normalen } = basis;
        const P = stoff.punkte, T = stoff.dreiecke, nP = P.length / 3;
        const { rand, nachbarn } = Hautmaske.randpunkte(T, nP);
        const locker = Hautmaske.lockereRandpunkte(P, rand, basis, eng);
        const ok = Hautmaske.dreieckeOhne(T, Hautmaske.ringeUm(locker, nachbarn, ringe));
        const umkreis = Hautmaske.dreieckeImUmkreis(T, nP, ok, nachbarn);
        const h = G.ZELLE_M;
        const gitter = G.punktgitter(P, h);
        // Nur Körperpunkte in der (erweiterten) Hülle des Stücks kommen in Frage.
        const huelle = G.huelle(P, suchweite);
        const grenze2 = suchweite * suchweite;
        for (let i = 0; i < maske.length; i++) {
            if (maske[i]) continue;
            const px = koerper[3 * i], py = koerper[3 * i + 1], pz = koerper[3 * i + 2];
            if (px < huelle[0] || px > huelle[3] || py < huelle[1] || py > huelle[4]
                || pz < huelle[2] || pz > huelle[5]) continue;
            const nah = G.naechsterPunkt(P, gitter, h, px, py, pz);
            if (!nah || nah.d2 > grenze2) continue;
            // Der Strahl entlang der Körpernormale, von `tiefe` in der Haut
            // bis `abstand` darüber, muss ein zulässiges Dreieck treffen.
            const nx = normalen[3 * i], ny = normalen[3 * i + 1], nz = normalen[3 * i + 2];
            const liste = umkreis(nah.j);
            for (let l = 0; l < liste.length; l++) {
                const k = liste[l];
                const t = G.strahlDreieck(P, T[k], T[k + 1], T[k + 2], px, py, pz, nx, ny, nz);
                if (t !== null && t >= -tiefe && t <= abstand) { maske[i] = 1; break; }
            }
        }
    }

    /**
     * Je Stoffpunkt die zulässigen Dreiecke bis zwei Nachbarringe weit —
     * gerechnet beim ersten Zugriff, dann gemerkt.
     *
     * WARUM DER STRAHL UND NICHT DER NÄCHSTE FLÄCHENPUNKT: Am Rand eines
     * anliegenden Stücks ist der nächste Flächenpunkt eines Körperpunkts
     * knapp jenseits der Kante die Kante selbst — 2 mm über der Haut, also
     * „darüber", und die Haut verschwände bis 25 mm über den Bund hinaus.
     * Der Strahl entlang der Normale trifft jenseits der Kante nichts.
     * Zwei Ringe reichen: Bei 8 mm Stoffnetz und 25 mm Abstand liegt der
     * Treffer höchstens 13 mm neben dem nächsten Stoffpunkt.
     */
    static dreieckeImUmkreis(T, n, ok, nachbarn) {
        const jePunkt = Hautmaske.dreieckeJePunkt(T, n, ok);
        const merker = new Array(n);
        return (j) => {
            if (merker[j]) return merker[j];
            const menge = new Set(jePunkt[j]);
            for (const w of nachbarn[j]) {
                for (const k of jePunkt[w]) menge.add(k);
                for (const x of nachbarn[w]) for (const k of jePunkt[x]) menge.add(k);
            }
            merker[j] = Array.from(menge);
            return merker[j];
        };
    }

    // ------------------------------------------------------------ Index

    /**
     * Der Index ohne Dreiecke, deren drei Ecken verdeckt sind. Die Gruppen
     * (`{start, count, materialIndex}`, in Indexeinträgen) folgen mit.
     * Ohne Gruppen gilt der ganze Index als eine.
     */
    static indexOhne(index, gruppen, maske) {
        const quellen = (gruppen && gruppen.length)
            ? gruppen
            : [{ start: 0, count: index.length, materialIndex: 0 }];
        const behalten = new Uint32Array(index.length);
        const neueGruppen = [];
        let lauf = 0, entfernt = 0;
        for (const g of quellen) {
            const start = lauf;
            const ende = Math.min(g.start + g.count, index.length);
            for (let k = g.start; k + 2 < ende; k += 3) {
                const a = index[k], b = index[k + 1], c = index[k + 2];
                if (maske[a] && maske[b] && maske[c]) { entfernt += 1; continue; }
                behalten[lauf] = a; behalten[lauf + 1] = b; behalten[lauf + 2] = c;
                lauf += 3;
            }
            neueGruppen.push({ start, count: lauf - start, materialIndex: g.materialIndex ?? 0 });
        }
        return { index: behalten.slice(0, lauf),
                 gruppen: (gruppen && gruppen.length) ? neueGruppen : [],
                 entfernt };
    }

    // ------------------------------------------------------------- Rand

    /** Die Punkte an offenen Kanten (Kante in genau einem Dreieck) und
     *  die Nachbarn je Punkt. */
    static randpunkte(T, n) {
        const zaehler = new Map(), nachbarn = Array.from({ length: n }, () => []);
        for (let k = 0; k + 2 < T.length; k += 3) {
            for (const [a, b] of [[T[k], T[k + 1]], [T[k + 1], T[k + 2]], [T[k + 2], T[k]]]) {
                const s = a < b ? a * n + b : b * n + a;
                const bisher = zaehler.get(s) || 0;
                zaehler.set(s, bisher + 1);
                if (!bisher) { nachbarn[a].push(b); nachbarn[b].push(a); }
            }
        }
        const rand = new Set();
        for (const [s, c] of zaehler) if (c === 1) { rand.add(Math.floor(s / n)); rand.add(s % n); }
        return { rand, nachbarn };
    }

    /**
     * Die Randpunkte, die LOCKER über der Haut liegen: mehr als `eng` entlang
     * der Normale des nächsten Körperpunkts — oder ohne Körper in Reichweite
     * (drei Zellen, 9 cm). Nur von ihnen aus bleibt ein Streifen frei.
     */
    static lockereRandpunkte(P, rand, basis, eng) {
        const { koerper, normalen, koerpergitter } = basis;
        const locker = new Set();
        for (const v of rand) {
            const qx = P[3 * v], qy = P[3 * v + 1], qz = P[3 * v + 2];
            const nah = G.naechsterPunkt(koerper, koerpergitter, G.ZELLE_M, qx, qy, qz);
            if (!nah) { locker.add(v); continue; }
            const i = nah.j;
            const d = (qx - koerper[3 * i]) * normalen[3 * i] + (qy - koerper[3 * i + 1]) * normalen[3 * i + 1]
                    + (qz - koerper[3 * i + 2]) * normalen[3 * i + 2];
            if (Math.abs(d) > eng) locker.add(v);
        }
        return locker;
    }

    /** `start` um `ringe` Nachbarringe erweitert. */
    static ringeUm(start, nachbarn, ringe) {
        let menge = new Set(start);
        for (let r = 0; r < ringe; r++) {
            const neu = new Set(menge);
            for (const v of menge) for (const w of nachbarn[v]) neu.add(w);
            menge = neu;
        }
        return menge;
    }

    /** Je Dreieck 1, wenn keine Ecke in `gesperrt` liegt. */
    static dreieckeOhne(T, gesperrt) {
        const ok = new Uint8Array(T.length / 3);
        for (let k = 0; k + 2 < T.length; k += 3) {
            ok[k / 3] = (gesperrt.has(T[k]) || gesperrt.has(T[k + 1]) || gesperrt.has(T[k + 2])) ? 0 : 1;
        }
        return ok;
    }

    /** Je Stoffpunkt die Anfänge der zulässigen Dreiecke, die ihn berühren. */
    static dreieckeJePunkt(T, n, ok) {
        const je = Array.from({ length: n }, () => []);
        for (let k = 0; k + 2 < T.length; k += 3) {
            if (!ok[k / 3]) continue;
            je[T[k]].push(k); je[T[k + 1]].push(k); je[T[k + 2]].push(k);
        }
        return je;
    }
}
